AddXHarnessCli in Helix SDK is not parallel-safe

dotnet/runtime makes parallel, parameterized submissions to Helix, mostly in GC stress testing different modes. This is a quick and effective way to submit the same job with different runtime parameters. For example, this block https://github.com/dotnet/runtime/blob/master/eng/pipelines/libraries/run-test-job.yml#L176-L184 causes the Helix submission to happen 7 times in parallel with different scenarios. This parallel launching is defined here: https://github.com/dotnet/runtime/blob/master/src/libraries/sendtohelix.proj#L23-L77

This does not work for cases where `<IncludeXHarnessCli>true</IncludeXHarnessCli>` is specified.

The `AddXHarnessCLI` target works by executing `dotnet tool install --tool-path /Users/directhex/Projects/runtime/artifacts/tmp/Release/Microsoft.DotNet.XHarness.CLI --version 1.0.0-prerelease.20403.2 --add-source https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json Microsoft.DotNet.XHarness.CLI` then specifying the tool's install directory as a Helix correlation payload - but fails when more than one Scenario is defined, because the `dotnet tool install` calls are not parallelisable.

Compare single-scenario:

```
  Done building Helix work items for scenario WasmTestOnBrowser. Work item count: 208
  You can invoke the tool using the following command: xharness
  Tool 'microsoft.dotnet.xharness.cli' (version '1.0.0-prerelease.20403.2') was successfully installed.
  Uploading payloads for Job on Ubuntu.1804.Amd64.Open...
```

to parallel-scenario:

```
  Done building Helix work items for scenario WasmTestOnBrowser. Work item count: 208
  Tool 'microsoft.dotnet.xharness.cli' (version '1.0.0-prerelease.20403.2') is already installed.
  Tool 'microsoft.dotnet.xharness.cli' failed to install. This failure may have been caused by:
  * You are attempting to install a preview release and did not use the --version option to specify the version.
  * A package by this name was found, but it was not a .NET tool.
  * The required NuGet feed cannot be accessed, perhaps because of an Internet connection problem.
  * You mistyped the name of the tool.
  For more reasons, including package naming enforcement, visit https://aka.ms/failure-installing-tool
/Users/directhex/Projects/runtime/.packages/microsoft.dotnet.helix.sdk/5.0.0-beta.20374.1/tools/xharness-runner/XHarnessRunner.targets(31,5): error MSB3073: The command "dotnet tool install --tool-path /Users/directhex/Projects/runtime/artifacts/tmp/Release/Microsoft.DotNet.XHarness.CLI --version 1.0.0-prerelease.20403.2 --add-source https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json Microsoft.DotNet.XHarness.CLI" exited with code 1. [/Users/directhex/Projects/runtime/src/libraries/sendtohelixhelp.proj]
##vso[task.logissue type=error;sourcepath=/Users/directhex/Projects/runtime/.packages/microsoft.dotnet.helix.sdk/5.0.0-beta.20374.1/tools/xharness-runner/XHarnessRunner.targets;linenumber=31;columnnumber=5;code=MSB3073;](NETCORE_ENGINEERING_TELEMETRY=Helix) The command "dotnet tool install --tool-path /Users/directhex/Projects/runtime/artifacts/tmp/Release/Microsoft.DotNet.XHarness.CLI --version 1.0.0-prerelease.20403.2 --add-source https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet-eng/nuget/v3/index.json Microsoft.DotNet.XHarness.CLI" exited with code 1.
  You can invoke the tool using the following command: xharness
  Tool 'microsoft.dotnet.xharness.cli' (version '1.0.0-prerelease.20403.2') was successfully installed.
  Uploading payloads for Job on Ubuntu.1804.Amd64.Open...
```

--- a/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets
+++ b/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets
@@ -25,20 +25,22 @@
     <!-- We install the tool locally on the build machine, then, installed as-is, we zip it up and send as a correlation payload -->
     <PropertyGroup>
       <_XHarnessLocalToolPath>$(ArtifactsTmpDir)Microsoft.DotNet.XHarness.CLI</_XHarnessLocalToolPath>
+      <_XHarnessCliPath>$(_XHarnessLocalToolPath)\.store\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)</_XHarnessCliPath>
     </PropertyGroup>
 
+    <!-- It can happen that the SDK is invoke in parallel and we have to make sure we don't install the tool twice -->
     <Exec Command="dotnet tool install --tool-path $(_XHarnessLocalToolPath) --version $(_XHarnessPackageVersion) --add-source $(XHarnessPackageSource) $(_XHarnessPackageName)"
           WorkingDirectory="$(ArtifactsTmpDir)"
-          Condition=" !Exists('$(_XHarnessLocalToolPath)\.store\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)') " />
+          IgnoreExitCode="true" />
 
     <!-- There are files we don't need extracted inside of the extracted tool dir (such as the original .nupkg) that blow up the size 3x, so we remove them -->
     <ItemGroup>
-      <_XHarnessExtraToolFiles Include="$(ArtifactsTmpDir)\Microsoft.DotNet.XHarness.CLI\.store\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)\*.*" />
+      <_XHarnessExtraToolFiles Include="$(_XHarnessCliPath)\*.*" />
     </ItemGroup>
     <Delete Files="@(_XHarnessExtraToolFiles)" />
 
     <ItemGroup>
-      <HelixCorrelationPayload Include="$(_XHarnessLocalToolPath)\.store\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)">
+      <HelixCorrelationPayload Include="$(_XHarnessCliPath)">
         <Destination>microsoft.dotnet.xharness.cli</Destination>
         <IncludeDirectoryName>true</IncludeDirectoryName>
       </HelixCorrelationPayload>

--- a/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets
+++ b/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets
@@ -27,9 +27,9 @@
       <_XHarnessLocalToolPath>$(ArtifactsTmpDir)Microsoft.DotNet.XHarness.CLI</_XHarnessLocalToolPath>
     </PropertyGroup>
 
-    <RemoveDir Directories="$(_XHarnessLocalToolPath)" />
     <Exec Command="dotnet tool install --tool-path $(_XHarnessLocalToolPath) --version $(_XHarnessPackageVersion) --add-source $(XHarnessPackageSource) $(_XHarnessPackageName)"
-          WorkingDirectory="$(ArtifactsTmpDir)" />
+          WorkingDirectory="$(ArtifactsTmpDir)"
+          Condition=" !Exists('$(_XHarnessLocalToolPath)\.store\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)') " />
 
     <!-- There are files we don't need extracted inside of the extracted tool dir (such as the original .nupkg) that blow up the size 3x, so we remove them -->
     <ItemGroup>

--- a/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets
+++ b/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/XHarnessRunner.targets
@@ -28,7 +28,7 @@
       <_XHarnessCliPath>$(_XHarnessLocalToolPath)\.store\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)\microsoft.dotnet.xharness.cli\$(_XHarnessPackageVersion)</_XHarnessCliPath>
     </PropertyGroup>
 
-    <!-- It can happen that the SDK is invoke in parallel and we have to make sure we don't install the tool twice -->
+    <!-- It can happen that the SDK is invoked in parallel and then it can fail when re-using the install location, so we ignore the exit code -->
     <Exec Command="dotnet tool install --tool-path $(_XHarnessLocalToolPath) --version $(_XHarnessPackageVersion) --add-source $(XHarnessPackageSource) $(_XHarnessPackageName)"
           WorkingDirectory="$(ArtifactsTmpDir)"
           IgnoreExitCode="true" />

